File: Python2/LeetCode/minimum-depth-of-binary-tree.py
__author__ = 'ryan'

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def minDepth(self, root):
        # recursive function
        # add one to min depth of children
        if root is None:
            return 0
        elif root.left is None and root.right is None:
            return 1
        elif root.left is None and root.right is not None:
            return 1 + self.minDepth(root.right)
        elif root.left is not None and root.right is None:
            return 1 + self.minDepth(root.left)
        else:
            return min(1 + self.minDepth(root.right), 1 + self.minDepth(root.left))

File: CPP/minimum_depth_of_binary_tree.cpp
#include <cstddef>
#include <algorithm>
#include <iostream>

/**
 * Definition for binary tree
*/
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int minDepth(TreeNode *root) {
        if (root == NULL){
            return 0;
        } else if (root->left == NULL && root->right == NULL){
            return 1;
        } else if (root->left == NULL && root->right != NULL){
            return 1 + minDepth(root->right);
        } else if (root->left != NULL && root->right == NULL){
            return 1 + minDepth(root->left);
        } else if (root->left != NULL && root->right != NULL){
            return 1 + std::min(minDepth(root->left), minDepth(root->right));
        }
    }
};

int main(){
	Solution sol;
	struct TreeNode node1(1);
	struct TreeNode node2(2,NULL,node1);
	std::cout << sol.minDepth(&node2);
}


